### PR TITLE
Maps/Instances: Implement handling of CMSG_SET_SAVED_INSTANCE_EXTEND

### PR DESCRIPTION
**Pull request by Treeston**
Title says it all, really.
- Add a new column to `character_instance` to store extension state
- Handle appropriate opcodes
- Adjust instance reset behavior to not reset instances that still have players bound to them
- Add a new `.debug raidreset` command because otherwise testing this would've just been painful.

### CODE REVIEW

**Comment by Ascathor** [comment]
This is awesome.

**Comment by jackpoz** [comment]
As a note, travis is complaining about
>"sql/updates/characters/2016_01_24_instanceextend.sql" is missing in the 'characters'.'updates' table.

and about a warning at /home/travis/build/TrinityCore/TrinityCore/src/server/game/Instances/InstanceSaveMgr.cpp:552
> redefinition of 'player'

File: src/server/game/Instances/InstanceSaveMgr.cpp [hunk 1]
```diff
@@ -516,10 +544,27 @@ void InstanceSaveManager::_ResetSave(InstanceSaveHashMap::iterator &itr)
     // do not allow UnbindInstance to automatically unload the InstanceSaves
     lock_instLists = true;
 
+    bool shouldDelete = true;
     InstanceSave::PlayerListType &pList = itr->second->m_playerList;
-    while (!pList.empty())
+    std::vector<Player*> temp; // list of expired binds that should be unbound
+    for (Player* player : pList)
     {
         Player* player = *(pList.begin());
```
Comment (jackpoz) [inline src/server/game/Instances/InstanceSaveMgr.cpp]: is this code currently executing the loop always using the first Player in the list ?

File: src/server/game/Instances/InstanceSaveMgr.cpp [hunk 2]
```diff
@@ -516,10 +544,27 @@ void InstanceSaveManager::_ResetSave(InstanceSaveHashMap::iterator &itr)
...
+        if (InstancePlayerBind* bind = player->GetBoundInstance(itr->second->GetMapId(), itr->second->GetDifficulty()))
+        {
+            assert(bind->save == itr->second);
```
Comment (jackpoz) [inline src/server/game/Instances/InstanceSaveMgr.cpp]: Please use the TC assert macro to ensure the check is valid in every build or remove it if it's not that much needed. Having different checks with different build modes makes crash logs more confusing.

File: src/server/game/Instances/InstanceSaveMgr.cpp (hunk 1 above)
Comment (Treeston) [inline src/server/game/Instances/InstanceSaveMgr.cpp]: That's old code that I must have missed while removing.

File: src/server/game/Instances/InstanceSaveMgr.cpp (hunk 2 above)
Comment (Shauren) [inline src/server/game/Instances/InstanceSaveMgr.cpp]: Under no circumstances you should ever use assert from <cassert> header - it does NOT generate a crash log on windows.

Comment (Treeston) [inline src/server/game/Instances/InstanceSaveMgr.cpp]: Yup, acknowledged.

File: src/server/scripts/Commands/cs_debug.cpp [hunk 3]
```diff
@@ -1442,6 +1443,33 @@ class debug_commandscript : public CommandScript
...
+        if (map <= 0)
+            return false;
+        MapEntry const* mEntry = sMapStore.LookupEntry(map);
+        if (!mEntry || !mEntry->IsRaid())
+            return false;
+        int32 difficulty = difficulty_str ? atoi(difficulty_str) : -1;
+        if (difficulty >= MAX_RAID_DIFFICULTY || difficulty < -1)
+            return false;
+
+        time_t t = time(nullptr);
```
Comment (tkrokli) [inline src/server/scripts/Commands/cs_debug.cpp]: `unused variable 't' [-Wunused-variable]` (compile fails, output from Travis)

Comment (Treeston) [inline src/server/scripts/Commands/cs_debug.cpp]: mm forgot to remove it after it stopped being needed

**Comment by ccrs** [comment]
travis loves u XD ````static bool HandleDebugRaidResetCommand(ChatHandler* handler, char c...````

File: src/server/game/Instances/InstanceSaveMgr.cpp [hunk 4]
```diff
@@ -439,6 +439,23 @@ void InstanceSaveManager::LoadResetTimes()
...
+        TC_LOG_ERROR("misc", "InstanceSaveManager::GetSubsequentResetTime: not valid difficulty or no reset delay for map %u", mapid);
+        return 0;
+    }
+    
+    time_t diff = sWorld->getIntConfig(CONFIG_INSTANCE_RESET_TIME_HOUR) * HOUR;
+    time_t period = uint32(((mapDiff->resetTime * sWorld->getRate(RATE_INSTANCE_RESET_TIME)) / DAY) * DAY);
+    if (period < DAY)
+        period = DAY;
+
+    return ((resetTime + MINUTE) / DAY * DAY) + period + diff;
```
Comment (treak) [inline src/server/game/Instances/InstanceSaveMgr.cpp]: wrong bracket? / DAY * DAY = 1

Comment (Treeston) [inline src/server/game/Instances/InstanceSaveMgr.cpp]: No, it isn't. Integer math,

Comment (treak) [inline src/server/game/Instances/InstanceSaveMgr.cpp]: ok sry :)

**Pull request merged**